File: src/main/java/com/HOIIVUtils/clauzewitz/data/focus/FocusTree.java
package com.HOIIVUtils.clauzewitz.data.focus;

import com.HOIIVUtils.clauzewitz.HOIIVFile;
import com.HOIIVUtils.clauzewitz.data.country.CountryTagsManager;
import com.HOIIVUtils.clauzewitz.script.*;
import com.HOIIVUtils.clauzewitz.localization.Localizable;
import com.HOIIVUtils.clauzewitz.data.country.CountryTag;

import javafx.collections.FXCollections;
import javafx.collections.MapChangeListener;
import javafx.collections.ObservableList;
import javafx.collections.ObservableMap;
import org.jetbrains.annotations.*;

import java.io.File;
import java.util.*;

/**
 * ALL of the FocusTree/FocusTrees
 * Localizable data: focus tree name. Each focus is its own localizable data.
 */
// todo extends file?
public class FocusTree extends StructuredPDX implements Localizable, Comparable<FocusTree>, Iterable<Focus> {
	private static final ObservableMap<File, FocusTree> focusTrees = FXCollections.observableHashMap();
	private static final ObservableList<FocusTree> focusTreesList = FXCollections.observableArrayList();
	static {
		focusTrees.addListener((MapChangeListener<File, FocusTree>) change -> {
			updateObservableValues();
			focusTreesList.sort(Comparator.naturalOrder());
		});
	}

	//private final ObservableMap<String, Focus> focuses;
	@NotNull public final ReferencePDXScript<CountryTag> country;
	@NotNull public final MultiPDXScript<Focus> focuses;
	@NotNull public final AbstractPDX<String> id;

	private ArrayList<String> focusIDList;
	@NotNull private File focus_file;
	// private Modifier countryModifier;
	// private boolean defaultFocus; // ! todo Do This
	// private Point continuousFocusPosition; // ! todo DO THIS

	public static FocusTree get(File focus_file) {
		if (!focusTrees.containsKey(focus_file)) {
			new FocusTree(focus_file);
		}
		return focusTrees.get(focus_file);
	}

	/**
	 * Instantiate focus tree from pre-existing focus tree file.
	 * 
	 * @param focus_file pre-existing focus tree.
	 */
	private FocusTree(@NotNull File focus_file) {
		super("focus_tree");
		this.focus_file = focus_file;

		/* pdxscript */
		id = new StringPDX("id");
		country = new ReferencePDXScript<>(CountryTagsManager::getCountryTags, CountryTag::get, "country");
		focuses = new MultiPDXScript<>(() -> new Focus(this), "focus");
		obj.addAll(childScripts());
		loadPDX(focus_file);

		FocusTree.add(this);
	}

	@Override
	protected Collection<? extends AbstractPDX<?>> childScripts() {
		return List.of(id, country, focuses);
	}

	public static ObservableList<FocusTree> observeFocusTrees() {
		return focusTreesList;
	}

	public static void read() {
		if (!HOIIVFile.mod_focus_folder.exists() || !HOIIVFile.mod_focus_folder.isDirectory()) {
			System.err.println("Focus folder does not exist or is not a directory.");
			return;
		}
		if (HOIIVFile.mod_focus_folder.listFiles() == null || HOIIVFile.mod_focus_folder.listFiles().length == 0) {
			System.err.println("No focuses found in " + HOIIVFile.mod_focus_folder);
			return;
		}

		for (File f : HOIIVFile.mod_focus_folder.listFiles()) {
			if (f.getName().endsWith(".txt")) {
				new FocusTree(f);
			}
		}
	}

//	private ArrayList<String> parse() {
//		if (this.focus_file == null) {
//			System.err.println(this + "File of focus tree not set.");
//			return null;
//		}
//		if (!this.focus_file.exists()) {
//			System.err.println(this + "Focus tree file does not exist.");
//			return null;
//		}
//
//		focusIDList = new ArrayList<>();
//		this.focuses.clear();
//
//		/* parser */
//		Parser focusTreeParser = new Parser(this.focus_file);
//		Node focusTreeNode;
//		try {
//			focusTreeNode = focusTreeParser.parse();
//			var l = focusTreeNode.filterName("focus_tree").toList();
//			if (l.isEmpty()) {
//				System.out.println("focus_tree filter yielded no results for " + focus_file);
//				return null;
//			}
//			focusTreeNode = l.get(0);
//		} catch (Exception e) {
//			System.err.println("Error parsing focus tree file: " + focus_file);
//			return focusIDList;
//			//throw new RuntimeException(e);
//		}
//
//		/* focus tree id */
//		try {
//			id = focusTreeNode.getValue("id").string();
//		} catch (NullPointerException e) {
//			System.err.println("Focus tree id not found, focus file: " + focus_file);
//			// todo throw exception?
//		}
//
//		/* country */
//		// todo should not be .findFirst("modifier"); need mofifier handling. but thats
//		// okay.
//		Node countryModifierExp = focusTreeNode.findFirst("country").findFirst("modifier");
//		if (countryModifierExp != null) {
//			try {
//				if (countryModifierExp.getValue("tag") != null)
//					country = new CountryTag(countryModifierExp.getValue("tag").string());
//				else
//					country = CountryTag.NULL_TAG;
//			} catch (NullPointerException e) {
//				System.out.println("Country modifier tag not found, focus file: " + focus_file);
//				throw new RuntimeException(e);
//			}
//		} else {
//			System.out.println("Country modifier not found, focus file: " + focus_file);
//		}
//
//		/* focuses */
//		List<Focus> focuses = getFocuses(focusTreeNode);
//		minX = focuses.stream().mapToInt(Focus::absoluteX).min().orElse(0);
//
//		return focusIDList;
//	}

//	@NotNull
//	private List<Focus> getFocuses(Node focusTreeNode) {
//		if (focusTreeNode.value().list() == null) {
//			System.err.println("Expected list of nodes, for focuses getter");
//			return new ArrayList<>();
//		}
//
//		List<Focus> focusList = new ArrayList<>();
//
//		List<Node> focusTreeNodes = focusTreeNode.filterName("focus").toList();
//		for (Node node : focusTreeNodes) {
//			Focus focus;
//			/* focus id */
//			try {
//				focus = new Focus(this, node);
//			} catch (DuplicateFocusException e) {
//				System.err.println(e.getMessage());
//				continue;
//			}
//			focusIDList.add(focus.id.get());
//			focuses.put(focus);
//			focusList.add(focus);
//		}
//
//		/* when all focuses init loaded */
//		checkPendingFocusReferences();
//
//		return focusList;
//	}

	private void checkPendingFocusReferences() {
//		List<Focus> resolvedReferences = new ArrayList<>();
//		// private final HashMap<String, Consumer<List<Node>>> pendingFocusReferences =
//		// new HashMap<>();
//
//		List<PendingFocusReferenceList> pendingFocusReferenceLists = focuses().parallelStream()
//				.map(Focus::getPendingFocusReferences).toList();
//
//		/* resolve references of focuses that exist */
//		for (var pendingFocusReferenceList : pendingFocusReferenceLists) {
//			var pendingFocusReferences = pendingFocusReferenceList.pendingFocusReferences;
//			List<String> referencesToRemove = pendingFocusReferences.stream()
//					.map(PendingFocusReference::id)
//					.filter(id -> focusIDList.contains(id)).toList();
//			referencesToRemove.forEach(pendingFocusReferenceList::resolve);
//			// todo temp want better warnings in future
//			/* unresolved references */
//			// pendingFocusReferences.forEach(pfr -> JOptionPane.showMessageDialog(null,
//			// "Undefined Focus reference invalid: " + pfr.id() + ", " +
//			// pfr.pendingActionMap().keySet()));
//			pendingFocusReferences.forEach(pfr -> System.out.println("Warning: [Focus.java] " +
//					"Undefined Focus reference invalid: " + pfr.id() + ", " + pfr.pendingActionMap().keySet()));
//		}

	}

	/**
	 * Lists last set of focuses
	 * 
	 * @return
	 */
	public ArrayList<String> listFocusIDs() {
		if (focusIDList == null) {
			return null; // bad :(
		}

		return focusIDList;
	}

	// public static ArrayList<String> list(File focus_file) throws IOException {
	// if(NationalFocuses.get(focus_file) focus_names == null) {
	// return FocusTree.find(focus_file);
	// }
	//
	// return focus_names;
	// }

	public CountryTag countryTag() {
		if (country.get() != null) {
			return country.get();
		} else {
			// idk :(
			return null;
		}
	}

	public File focusFile() {
		return focus_file;
	}

	@Override
	public boolean equals(Object other) {
		if (other == null)
			return false;

		if (other.getClass() == this.getClass()) {
			return this.focus_file == ((FocusTree) other).focus_file;
		}

		return false;
	}

	@Override
	public String toString() {
		if (id.get() != null) {
			return id.get();
		}
		if (country.get() != null) {
			return country.get().toString();
		}
		return super.toString();
	}

	public @NotNull HashSet<Focus> focuses() {
		var list = this.focuses.get();
		if (list == null) return new HashSet<>();
		return new HashSet<>(list);
	}

	public int minX() {
		return focuses().stream().map(Focus::absoluteX).min(Integer::compareTo).orElse(0);
	}

	public static HashMap<File, FocusTree> add(FocusTree focusTree) {
		focusTrees.put(focusTree.focus_file, focusTree);
		return new HashMap<>(focusTrees);
	}

	public static FocusTree[] listFocusTrees() {
		return focusTrees.values().toArray(new FocusTree[0]);
	}

	/**
	 * Returns focus tree corresponding to the tag, if it exists
	 * 
	 * @param tag
	 * @return The focus tree, or null if could not be found/not yet created.
	 */
	public static FocusTree get(CountryTag tag) {
//		return focusTrees.get(tag);
		return focusTrees.values().stream().filter(focusTree -> focusTree.country.objEquals(tag)).findFirst().orElse(null);
	}

	public static FocusTree getdankwizardisfrench(CountryTag tag) {
		for (FocusTree tree : listFocusTrees()) {
			assert tree.country.get() != null;
			if (tree.country.objEquals(tag)) {
				return tree;
			}
		}

		return null;
	}

	public ObservableList<Focus> listFocuses() {
		ObservableList<Focus> focusList = FXCollections.observableArrayList();
		focusList.addAll(focuses());
		return focusList;
	}

	private static void updateObservableValues() {
		focusTreesList.setAll(focusTrees.values());
	}

	@Override
	public int compareTo(@NotNull FocusTree o) {
		int c = 0;
		if (this.country.get() != null && o.country.get() != null) {
			c = this.country.get().compareTo(o.country.get());
		}
		int d = 0;
		if (this.id.get() != null && o.id.get() != null) {
			d = this.id.get().compareTo(o.id.get());
		}
		return c == 0 ? d : c;
	}

	@NotNull
	@Override
	public Iterator<Focus> iterator() {
		return focuses().iterator();
	}

	@Override
	public @NotNull Map<Property, String> getLocalizableProperties() {
		// lets us map null if we use hashmap instead of generic of() method
		HashMap<Property, String> properties = new HashMap<>();
		properties.put(Property.NAME, id.get());
		return properties;
	}

	/**
	 * Get the localizable group of this focus tree, which is the list of focuses.
	 * @return
	 */
	@Override
	public @NotNull Collection<? extends Localizable> getLocalizableGroup() {
		return focuses();
	}

	@Override
	public boolean objEquals(PDXScript<?> other) {
		if (other instanceof FocusTree) {
			return this.equals(other);
		}
		return false;
	}
}

File: src/main/scala/com/hoi4utils/clausewitz/data/focus/FocusTree.scala
package com.hoi4utils.clausewitz.data.focus

import language.experimental.namedTuples

import com.hoi4utils.clausewitz.data.country.{CountryTag, CountryTagsManager}
import com.hoi4utils.clausewitz.data.focus.FocusTree.focusTreeFileMap
import com.hoi4utils.clausewitz.localization.*
import com.hoi4utils.clausewitz.script.*
import javafx.collections.{FXCollections, ObservableList}
import org.apache.logging.log4j.{LogManager, Logger}
import com.hoi4utils.clausewitz.{HOIIVUtils, HOIIVFiles}
import org.jetbrains.annotations.*

import java.io.File
import scala.collection.mutable
import scala.collection.mutable.{HashMap, ListBuffer}
import scala.jdk.javaapi.CollectionConverters

/**
 * ALL the FocusTree/FocusTrees
 * Localizable data: focus tree name. Each focus is its own localizable data.
 */
// todo extends file?
object FocusTree {
  val LOGGER: Logger = LogManager.getLogger(classOf[FocusTree])
  private val focusTreeFileMap = new mutable.HashMap[File, FocusTree]()
  private val focusTrees = new ListBuffer[FocusTree]()

  def get(focus_file: File): Option[FocusTree] = {
    if (focus_file == null) return None
    if (!focusTreeFileMap.contains(focus_file)) new FocusTree(focus_file)
    focusTreeFileMap.get(focus_file)
  }

  def observeFocusTrees: ObservableList[FocusTree] = {
    FXCollections.observableArrayList(CollectionConverters.asJava(focusTrees))
  }

  /**
   * Reads all focus trees from the focus trees folder, creating FocusTree instances for each.
   */
  def read(): Boolean = {
    if (!HOIIVFiles.Mod.focus_folder.exists || !HOIIVFiles.Mod.focus_folder.isDirectory) {
      LOGGER.fatal(s"In ${this.getClass.getSimpleName} - ${HOIIVFiles.Mod.focus_folder} is not a directory, or it does not exist.")
      false
    } else if (HOIIVFiles.Mod.focus_folder.listFiles == null || HOIIVFiles.Mod.focus_folder.listFiles.length == 0) {
      LOGGER.warn(s"No focuses found in ${HOIIVFiles.Mod.focus_folder}")
      false
    } else {
      LOGGER.info("Reading focus trees from " + HOIIVFiles.Mod.focus_folder)

      // create focus trees from files
      HOIIVFiles.Mod.focus_folder.listFiles().filter(_.getName.endsWith(".txt")).foreach { f =>
        new FocusTree(f)
      }
      true
    }
  }

  /**
   * Clears all focus trees and any other relevant values.
   */
  def clear(): Unit = {
    focusTrees.clear()
    focusTreeFileMap.clear()
  }

  /**
   * Adds a focus tree to the list of focus trees.
   * @param focusTree the focus tree to add
   * @return the updated list of focus trees
   */
  def add(focusTree: FocusTree): Iterable[FocusTree] = {
    focusTrees += focusTree
    focusTree.focusFile match {
      case Some(file) => focusTreeFileMap.put(file, focusTree)
      case None =>
    }
    focusTrees
  }

  def listFocusTrees: Iterable[FocusTree] = focusTrees

  /**
   * Returns focus tree corresponding to the tag, if it exists
   *
   * @param tag The country tag
   * @return The focus tree, or null if could not be found/not yet created.
   */
  def get(tag: CountryTag): FocusTree = {
    //focusTrees.values.stream.filter((focusTree: FocusTree) => focusTree.country.nodeEquals(tag)).findFirst.orElse(null)
    for (tree <- listFocusTrees) {
      //if (tree.country.equals(tag)) return tree
      val countryTag = tree.countryTag
      countryTag match {
        case Some(t) => if (tag.equals(t.tag)) return tree
        case None => 
      }
    }
    null
  }
}

/**
 * Represents a focus tree, which is a collection of focuses.
 *
 * @note Do not create instances of this class directly, unless a few focus tree is being created or loaded.
 *       Use FocusTree.get(File) instead.
 */
class FocusTree
  extends StructuredPDX("focus_tree") with Localizable with Comparable[FocusTree] with Iterable[Focus] with PDXFile {
  /* pdxscript */
  //final var country = new ReferencePDX[CountryTag](() => CountryTag.toList, tag => Some(tag.get), "country")
  final var country = new FocusTreeCountryPDX
  final var focuses = new MultiPDX[Focus](None, Some(() => new Focus(this)), "focus")
  final var id = new StringPDX("id")
  // private boolean defaultFocus; // ! todo Do This
  // private Point continuousFocusPosition; // ! todo DO THIS

  private var _focusFile: Option[File] = None

  /* default */
  FocusTree.add(this)

  @throws[IllegalArgumentException]
  def this(focus_file: File) = {
    this()
    if (!focus_file.exists) {
      LOGGER.fatal(s"Focus tree file does not exist: $focus_file")
      throw new IllegalArgumentException(s"File does not exist: $focus_file")
    }

    loadPDX(focus_file)
    setFile(focus_file)
  }

  // todo: add default, continuous focus position
  override protected def childScripts: mutable.Iterable[? <: PDXScript[?]] = {
    ListBuffer(id, country, focuses)
  }

  /**
   * List of all focus IDs in this focus tree.
   *
   * @return list containing each focus ID
   */
  def listFocusIDs: Seq[String] = Seq.from(focuses.flatMap(_.id.value))
  
  def focusFile: Option[File] = _focusFile

  override def toString: String = {
    id.value.orElse(country.value.map(_.toString)).getOrElse {
      _focusFile match {
        case Some(file) if file.exists() => s"[Unknown ID: ${file.getName}]"
        case _ => "[Unknown]"
      }
    }
  }

  /**
   * Get the minimum X coordinate of all focuses in this focus tree.
   * @return the minimum absolute X coordinate of all focuses
   */
  def minX: Int = focuses.map(_.absoluteX).minOption.getOrElse(0)

  def listFocuses: List[Focus] = focuses.toList

  def addNewFocus(f: Focus): Unit = {
    focuses += f
  }

  /**
   * Get the next temporary focus ID.
   * @return the next temporary focus ID
   *
   * @note for Java compatibility -_- (as opposed to being able to just use Scala's default parameter)
   */
  def nextTempFocusID(): String = {
    nextTempFocusID(focuses.size)
  }

  /**
   * Get the next temporary focus ID.
   * @param lastIntID the last integer ID (should default to size of the focus tree)
   * @return the next temporary focus ID
   */
  private def nextTempFocusID(lastIntID: Int): String = {
    val id = "focus_" + (lastIntID + 1)
    if (focuses.exists(_.id.value.contains(id))) return nextTempFocusID(lastIntID)
    id
  }

  def setID(s: String): Unit = {
    this.id.set(s)
  }

  def setCountryTag(tag: CountryTag): Unit = {
    country.modifier.length match {
      case 0 => //country.modifier.add(new country.TagModifierPDX)
      case _ => country.modifier.head.tag.set(tag)
    }
  }

  def setFile(file: File): Unit = {
    _focusFile = Some(file)
    _focusFile.foreach(file => focusTreeFileMap.put(file, this))
  }
  
  override def getFile: Option[File] = {
    _focusFile
  }

  def countryTag: Option[CountryTag] = {
    country.modifier.length match {
      case 0 => None
      case _ => country.modifier.head.tag.value
    }
  }

  override def compareTo(o: FocusTree): Int = {
    (this.countryTag, this.id.value) match {
      case (Some(countryTag), Some(id)) =>
        (o.countryTag, o.id.value) match {
          case (Some(otherCountryTag), Some(otherID)) =>
            val c = countryTag.compareTo(otherCountryTag)
            if (c == 0) id.compareTo(otherID) else c
          case _ => 0
        }
      case _ => 0
    }
  }

  override def iterator: Iterator[Focus] = focuses.iterator

  override def getLocalizableProperties: mutable.HashMap[Property, String] = {
    // lets us map null if we use hashmap instead of generic of() method
    val properties = new mutable.HashMap[Property, String]
    id.value match {
      case Some(id) => properties.put(Property.NAME, id)
      case None => //properties.put(Property.NAME, null)
    }
    properties
  }

  /**
   * Get the localizable group of this focus tree, which is the list of focuses.
   *
   * @return
   */
  override def getLocalizableGroup: Iterable[? <: Localizable] = focuses

  override def equals(other: PDXScript[?]): Boolean = {
    if (other.isInstanceOf[FocusTree]) return this == other
    false
  }
  
  class FocusTreeCountryPDX extends StructuredPDX("country") {
    final val base = new DoublePDX("base")
    final val factor = new DoublePDX("factor")
    final val add = new DoublePDX("add")
    final val modifier = new MultiPDX[TagModifierPDX](None, Some(() => new TagModifierPDX), "modifier")

    override protected def childScripts: mutable.Iterable[? <: PDXScript[?]] = ListBuffer(base, factor, add, modifier)

    override def getPDXTypeName: String = "AI Willingness"

    class TagModifierPDX extends StructuredPDX("modifier") {
      final val base = new DoublePDX("base")
      final val factor = new DoublePDX("factor")
      final val add = new DoublePDX("add")
      final val tag = new ReferencePDX[CountryTag](() => CountryTag.toList, tag => Some(tag.get), "country")

      override protected def childScripts: mutable.Iterable[? <: PDXScript[?]] = ListBuffer(base, factor, add, tag)

      override def getPDXTypeName: String = "Modifier"
    }
  }
}